src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultDocumentVersionCache.cs

```diff
@@ -36,27 +33,34 @@ public override void TrackDocumentVersion(IDocumentSnapshot documentSnapshot, in
             throw new ArgumentNullException(nameof(documentSnapshot));
         }
 
-        _dispatcher.AssertDispatcherThread();
-
         var filePath = documentSnapshot.FilePath.AssumeNotNull();
+        using var upgradeableReadLock = _lock.EnterUpgradeAbleReadLock();
+        TrackDocumentVersion(documentSnapshot, version, filePath, upgradeableReadLock);
+    }
 
-        if (!DocumentLookup.TryGetValue(filePath, out var documentEntries))
+    private void TrackDocumentVersion(IDocumentSnapshot documentSnapshot, int version, string filePath, ReadWriterLocker.UpgradeableReadLock upgradeableReadLock)
+    {
+        // Need to ensure the write lock covers all uses of documentEntries, not just DocumentLookup
+        using (upgradeableReadLock.EnterWriteLock())
```

Comment:
This might just be me not understanding locks properly, but is there a reason why we're using `using ()` syntax instead of `using var _` syntax as we do in most other parts of this file? Is the intention to just be more explicit about the scope of the write lock?

Reply:
For write locks we're being a bit more explicit about where it is held, since they block all other operations.